src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+        let lognpnf = (n_p_f64 / n_f_f64).log2();
+        let lambda_rel = params.lambda_rel as f64;
+        let logrel = lambda_rel.log2();
+        let lambda_sec = params.lambda_sec as f64;
+        let loge = LOG2_E as f64;
+
+        let u_f64 = ((lambda_sec + logrel + 5.0 - loge.log2()) / lognpnf).ceil();
+        let u = u_f64 as u64;
+
+        let ratio = 9.0 * n_p_f64 * loge / ((17.0 * u_f64).powi(2));
+        let s1 = ratio - 7.0;
+        let s2 = ratio - 2.0;
+
+        if s1 < 1.0 || s2 < 1.0 {
+            // Small case, ie n_p <= λ^2
+            let ln12 = (12f64).ln();
+            let d = (32.0 * ln12 * u_f64).ceil();
+            return Setup {
+                n_p: params.n_p,
+                u,
+                r: params.lambda_rel as u64,
+                d: d as u64,
+                q: 2.0 * ln12 / d,
+                b: (8.0 * (u_f64 + 1.0) * d / ln12).floor() as u64,
+            };
+        }
+        let lambda_rel2 = lambda_rel.min(s2);
+        if u_f64 < lambda_rel2 {
+            // Case 3, Theorem 14, ie  n_p >= λ^3
+            let d = (16.0 * u_f64 * (lambda_rel2 + 2.0) / loge).ceil();
+            debug_assert!(n_p_f64 >= d * d * loge / (9.0 * (lambda_rel2 + 2.0)));
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel2).ceil() as u64,
+                d: d as u64,
+                q: 2.0 * (lambda_rel2 + 2.0) / (d * loge),
+                b: (((lambda_rel2 + 2.0 + u_f64.log2()) / (lambda_rel2 + 2.0))
+                    * (3.0 * u_f64 * d / 4.0)
+                    + d
+                    + u_f64)
+                    .floor() as u64,
+            }
+        } else {
+            // Case 2, Theorem 13, ie λ^3 > n_p > λ^2
+            let lambda_rel1 = lambda_rel.min(s1);
+            let lbar = (lambda_rel1 + 7.0) / loge;
+            let d = (16.0 * u_f64 * lbar).ceil();
+            debug_assert!(n_p_f64 >= d * d / (9.0 * lbar));
+
+            let w = compute_w(u_f64, lambda_rel1);
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel1).ceil() as u64,
+                d: d as u64,
+                q: 2.0 * lbar / d,
+                b: ((w * lbar / d + 1.0)
+                    * (2.0 * u_f64 * w * lbar / n_p_f64 + 7.0 * u_f64 / w).exp()
+                    * d
+                    * u_f64
+                    + d)
+                    .floor() as u64,
+            }
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    // Round candidate tuple
+    s_list: Vec<Element>,
+    /// Round candidate hash
+    h: Hash,
+    /// Round candidate hash mapped to [1, n_p]
+    h_u64: u64,
+    /// Approximate size of set Sp to lower bound
+    n_p: u64,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
```

Comment:
@rrtoledo - there might be a way to improve the memory usage in this function as follows: instead of creating the intermediate `data` (a structure that is `Vec<Vec<>>`) and use memory for it, you can use the `hasher` directly and call `hasher.update()` in-place. That will save you memory - something like this: 
```
fn h1(input: &[Vec<u8>], n_p: u64) -> (Hash, u64) {
    let mut hasher = Blake2b::new();
    hasher.update(b"Telescope-H1");
    for i in input {
        hasher.update(i);
    }
    let digest: Hash = hasher.finalize().try_into().unwrap();
    (digest, utils::sample_uniform(&digest, n_p))
}
```